src/utils/avatar.js

```diff
@@ -1,9 +1,257 @@
+
+    return new GravatarURL(standardSizeUrl);
+  }
+
+  static ORIGIN = 'https://secure.gravatar.com';
+
+  /**
+   * Standard URL from which to generate others. PRIVATE.
+   *
+   * May be a string if the instance was constructed at rehydrate
+   * time, when URL validation is unnecessary.
+   */
+  _standardUrl: string | URL;
+
+  /**
+   * PRIVATE: Make an instance from already-validated data.
+   *
+   * Not part of the public interface; use the static methods instead.
+   *
+   * It's private because we need a path to constructing an instance
+   * without constructing URL objects, which takes more time than is
+   * acceptable when we can avoid it, e.g., during rehydration.
+   * Constructing URL objects is a necessary part of validating data
+   * from the server, but we only need to validate the data once, when
+   * it's first received.
+   */
+  constructor(standardUrl: string | URL) {
+    super();
+    this._standardUrl = standardUrl;
+  }
+
+  /**
+   * Get a URL object for the given size.
+   *
+   * `sizePhysicalPx` must be an integer. (Gravatar doesn't advertise
+   * the ability to specify noninteger values for the size.)
+   */
+  get(sizePhysicalPx: number): URL {
+    // `this._standardUrl` may have begun its life as a string, to
+    // avoid computing a URL object during rehydration
+    if (typeof this._standardUrl === 'string') {
+      this._standardUrl = new URL(this._standardUrl);
+    }
+
+    // Make a new URL to mutate
+    // $FlowFixMe - https://github.com/zulip/zulip-mobile/pull/4230#discussion_r512351202
+    const result: URL = new URL(this._standardUrl);
+    result.searchParams.set('s', sizePhysicalPx.toString());
+    return result;
+  }
+}
+
+/**
+ * An avatar that was uploaded to the Zulip server.
+ *
+ * There are two size options; if `sizePhysicalPx` is greater than
+ * 100, medium is chosen:
+ *  * default: 100x100
+ *  * medium: 500x500
+ */
+export class UploadedAvatarURL extends AvatarURL {
+  /**
+   * Serialize to a special string; reversible with `deserialize`.
+   */
+  static serialize(instance: UploadedAvatarURL): string {
+    return instance._standardUrl instanceof URL
+      ? instance._standardUrl.toString()
+      : instance._standardUrl;
+  }
+
+  /**
+   * Use a special string from `serialize` to make a new instance.
+   */
+  static deserialize(serialized: string): UploadedAvatarURL {
+    return new UploadedAvatarURL(serialized);
+  }
+
+  /**
+   * Construct from raw server data, or throw an error.
+   *
+   * Expects a relative URL plus the realm for a local upload;
+   * otherwise, an absolute URL of the avatar on the S3 backend.
+   */
+  static validateAndConstructInstance(args: {|
+    realm: URL,
+    absoluteOrRelativeUrl: string,
+  |}): UploadedAvatarURL {
+    const { realm, absoluteOrRelativeUrl } = args;
+    // If `absoluteOrRelativeUrl` is absolute, the second argument
+    // is ignored.
+    return new UploadedAvatarURL(new URL(absoluteOrRelativeUrl, realm.origin));
```

Comment:
OK, and on doing that comparison with the help of the "switch from old to new implementation" pivot commit, I found one small difference in the logic!

The old implementation applies the possibly-relative avatar URL to the realm like so:
```js
new URL(avatarUrl, realm).toString()
```
while the new implementation uses `realm.origin` instead of `realm`.

I think this one shouldn't actually affect our behavior -- in particular, if the avatar URL is relative then it should have a leading `/`, so `new URL` will ignore the path (and everything after it) in the base URL if there is any. And we don't support nontrivial path (or query or fragment) in realm URLs anyway -- the one bit of scope for variation there is we haven't always applied consistent conventions about whether there's a trailing slash.

But the old code agrees with other places in our codebase, and the new code doesn't, so probably best to keep it uniform. That'd reduce surprises if it does later turn out there can be a difference, and in any case it makes one less thing to sort out when understanding the code.

Reply:
Thanks for catching this!